docs/design/ovs-pipeline.md

```diff
@@ -1,1197 +1,1852 @@
+[EndpointDNAT] to which the packets are forwarded and in which DNAT will be performed. `RemoteEndpointRegMark` is loaded
+for remote Endpoints, like the bucket with `bucket_id` 1 in this group.
 
-If the `conjunction` action is matched, packets are "allowed" or "dropped"
-based on the `action` field of the policy rule. If allowed, they follow a similar
-path as described in the following [EgressRuleTable] section.
+The third group with `group_id` 11 is the destination of packets matched by flow 6, designed for a Service that has
+Endpoints and is configured with session affinity. The group closely resembles the group with `group_id` 10, except that
+the destination of the packets is table [ServiceLB], rather than table [EndpointDNAT]. After being sent back to table
+[ServiceLB], they will be matched by flow 7.
 
-Unlike the default of K8s NetworkPolicies, Antrea-native policies have no such
-default rules. Hence, they are evaluated as-is, and there is no need for a
-AntreaPolicyEgressDefaultTable.
+### EndpointDNAT
 
-### EgressRuleTable (50)
+The table implements DNAT for Service connection after Endpoint selection is performed in table [ServiceLB].
 
-For this table, you will need to keep mind the Network Policy
-[specification](#network-policy-implementation) that we are using. We have 2
-Pods running on the same Node, with IP addresses 10.10.1.2 to 10.10.1.3. They
-are allowed to talk to each other using TCP on port 80, but nothing else.
-
-This table is used to implement the egress rules across all Network Policies. If
-you dump the flows for this table, you should see something like this:
+If you dump the flows of this table, you may see the following::
 
 ```text
-1. table=50, priority=210,ct_state=-new+est,ip actions=goto_table:70
-2. table=50, priority=200,ip,nw_src=10.10.1.2 actions=conjunction(2,1/3)
-3. table=50, priority=200,ip,nw_src=10.10.1.3 actions=conjunction(2,1/3)
-4. table=50, priority=200,ip,nw_dst=10.10.1.2 actions=conjunction(2,2/3)
-5. table=50, priority=200,ip,nw_dst=10.10.1.3 actions=conjunction(2,2/3)
-6. table=50, priority=200,tcp,tp_dst=80 actions=conjunction(2,3/3)
-7. table=50, priority=190,conj_id=2,ip actions=load:0x2->NXM_NX_REG5[],ct(commit,table=61,zone=65520,exec(load:0x2->NXM_NX_CT_LABEL[32..63]))
-8. table=50, priority=0 actions=goto_table:60
+1. table=EndpointDNAT, priority=200,reg0=0x4000/0x4000 actions=controller(reason=no_match,id=62373,userdata=04)
+2. table=EndpointDNAT, priority=200,tcp,reg3=0xa0a0018,reg4=0x20050/0x7ffff actions=ct(commit,table=AntreaPolicyEgressRule,zone=65520,nat(dst=10.10.0.24:80),exec(set_field:0x10/0x10->ct_mark,move:NXM_NX_REG0[0..3]->NXM_NX_CT_MARK[0..3]))
+3. table=EndpointDNAT, priority=200,tcp,reg3=0xa0a0106,reg4=0x20050/0x7ffff actions=ct(commit,table=AntreaPolicyEgressRule,zone=65520,nat(dst=10.10.1.6:80),exec(set_field:0x10/0x10->ct_mark,move:NXM_NX_REG0[0..3]->NXM_NX_CT_MARK[0..3]))
```

Comment:
quick question out of curiosity: OVS does not let us invoke the `nat` action with a `dst` value stored in a register? I didn't remember that we had individual flows for each endpoint her.

cc @wenyingd @tnqn

Reply:
The doc says:

> nat(type=addrs[:ports][,flag]...)
>
> The src and dst options take the following arguments:
> 
> addrs
> The IP address addr or range addr1-addr2 from which the translated address should be selected. If only one address is given, then that address will always be selected, otherwise the address selection can be informed by the optional persistent flag as described below. Either IPv4 or IPv6 addresses can be provided, but both addresses must be of the same type, and the datapath behavior is undefined in case of providing IPv4 address range for an IPv6 packet, or IPv6 address range for an IPv4 packet. IPv6 addresses must be bracketed with [ and ] if a port range is also given.

So the answer should be no.